File: services/user-service/src/test/java/com/visor/school/userservice/model/UserTest.java
package com.visor.school.userservice.model;

import org.junit.jupiter.api.Test;

import java.util.HashSet;
import java.util.Set;

import static org.junit.jupiter.api.Assertions.*;

class UserTest {

    @Test
    void shouldCreateUserWithRequiredFields() {
        Set<UserRole> roles = new HashSet<>();
        roles.add(UserRole.TEACHER);
        
        User user = new User(
            "keycloak-123",
            "test@example.com",
            roles,
            "John",
            "Doe"
        );

        // id is nullable and will be null until persisted by JPA
        assertNull(user.getId());
        assertEquals("keycloak-123", user.getKeycloakId());
        assertEquals("test@example.com", user.getEmail());
        assertTrue(user.getRoles().contains(UserRole.TEACHER));
        assertEquals("John", user.getFirstName());
        assertEquals("Doe", user.getLastName());
        assertFalse(user.isEmailVerified());
        assertEquals(AccountStatus.ACTIVE, user.getAccountStatus());
    }

    @Test
    void shouldRequireKeycloakId() {
        // This test validates that keycloakId is required
        // In actual implementation, this would be enforced by JPA @Column(nullable = false)
        Set<UserRole> roles = new HashSet<>();
        roles.add(UserRole.STUDENT);
        
        User user = new User(
            "keycloak-123",
            "test@example.com",
            roles,
            "Jane",
            "Smith"
        );

        assertNotNull(user.getKeycloakId());
        assertFalse(user.getKeycloakId().isBlank());
    }

    @Test
    void shouldUpdateLastLoginTimestamp() throws InterruptedException {
        Set<UserRole> roles = new HashSet<>();
        roles.add(UserRole.TEACHER);
        
        User user = new User(
            "keycloak-123",
            "test@example.com",
            roles,
            "John",
            "Doe"
        );

        var initialUpdatedAt = user.getUpdatedAt();
        var initialLastLoginAt = user.getLastLoginAt();

        Thread.sleep(10); // Small delay to ensure timestamp difference
        user.updateLastLogin();

        assertNotNull(user.getLastLoginAt());
        assertNotEquals(initialLastLoginAt, user.getLastLoginAt());
        assertTrue(user.getUpdatedAt().isAfter(initialUpdatedAt));
    }

    @Test
    void shouldVerifyEmail() throws InterruptedException {
        Set<UserRole> roles = new HashSet<>();
        roles.add(UserRole.TEACHER);
        
        User user = new User(
            "keycloak-123",
            "test@example.com",
            roles,
            "John",
            "Doe"
        );

        assertFalse(user.isEmailVerified());
        var initialUpdatedAt = user.getUpdatedAt();

        Thread.sleep(10);
        user.verifyEmail();

        assertTrue(user.isEmailVerified());
        assertTrue(user.getUpdatedAt().isAfter(initialUpdatedAt));
    }

    @Test
    void shouldUpdateAccountStatus() throws InterruptedException {
        Set<UserRole> roles = new HashSet<>();
        roles.add(UserRole.TEACHER);
        
        User user = new User(
            "keycloak-123",
            "test@example.com",
            roles,
            "John",
            "Doe"
        );

        assertEquals(AccountStatus.ACTIVE, user.getAccountStatus());
        var initialUpdatedAt = user.getUpdatedAt();

        Thread.sleep(10);
        user.updateStatus(AccountStatus.SUSPENDED);

        assertEquals(AccountStatus.SUSPENDED, user.getAccountStatus());
        assertTrue(user.getUpdatedAt().isAfter(initialUpdatedAt));
    }
}

File: services/user-service/src/test/kotlin/com/visor/school/userservice/model/UserTest.kt
package com.visor.school.userservice.model

import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import java.util.UUID

class UserTest {

    @Test
    fun `should create user with required fields`() {
        val user = User(
            keycloakId = "keycloak-123",
            email = "test@example.com",
            roles = mutableSetOf(UserRole.TEACHER),
            firstName = "John",
            lastName = "Doe"
        )

        // id is nullable and will be null until persisted by JPA
        assertNull(user.id)
        assertEquals("keycloak-123", user.keycloakId)
        assertEquals("test@example.com", user.email)
        assertTrue(user.roles.contains(UserRole.TEACHER))
        assertEquals("John", user.firstName)
        assertEquals("Doe", user.lastName)
        assertFalse(user.emailVerified)
        assertEquals(AccountStatus.ACTIVE, user.accountStatus)
    }

    @Test
    fun `should require keycloakId`() {
        // This test validates that keycloakId is required
        // In actual implementation, this would be enforced by JPA @Column(nullable = false)
        val user = User(
            keycloakId = "keycloak-123",
            email = "test@example.com",
            roles = mutableSetOf(UserRole.STUDENT),
            firstName = "Jane",
            lastName = "Smith"
        )

        assertNotNull(user.keycloakId)
        assertFalse(user.keycloakId.isBlank())
    }

    @Test
    fun `should update last login timestamp`() {
        val user = User(
            keycloakId = "keycloak-123",
            email = "test@example.com",
            roles = mutableSetOf(UserRole.TEACHER),
            firstName = "John",
            lastName = "Doe"
        )

        val initialUpdatedAt = user.updatedAt
        val initialLastLoginAt = user.lastLoginAt

        Thread.sleep(10) // Small delay to ensure timestamp difference
        user.updateLastLogin()

        assertNotNull(user.lastLoginAt)
        assertNotEquals(initialLastLoginAt, user.lastLoginAt)
        assertTrue(user.updatedAt.isAfter(initialUpdatedAt))
    }

    @Test
    fun `should verify email`() {
        val user = User(
            keycloakId = "keycloak-123",
            email = "test@example.com",
            roles = mutableSetOf(UserRole.TEACHER),
            firstName = "John",
            lastName = "Doe"
        )

        assertFalse(user.emailVerified)
        val initialUpdatedAt = user.updatedAt

        Thread.sleep(10)
        user.verifyEmail()

        assertTrue(user.emailVerified)
        assertTrue(user.updatedAt.isAfter(initialUpdatedAt))
    }

    @Test
    fun `should update account status`() {
        val user = User(
            keycloakId = "keycloak-123",
            email = "test@example.com",
            roles = mutableSetOf(UserRole.TEACHER),
            firstName = "John",
            lastName = "Doe"
        )

        assertEquals(AccountStatus.ACTIVE, user.accountStatus)
        val initialUpdatedAt = user.updatedAt

        Thread.sleep(10)
        user.updateStatus(AccountStatus.SUSPENDED)

        assertEquals(AccountStatus.SUSPENDED, user.accountStatus)
        assertTrue(user.updatedAt.isAfter(initialUpdatedAt))
    }
}

